### PR TITLE
feat: Create wrapAllProxyETH (DEV-1144)

### PR DESCRIPTION
**Pull request by TheMj0ln1r**
## Summary by CodeRabbit

* **New Features**
  * Introduced ETH wrapping functionality enabling conversion of ETH holdings into WETH directly through the proxy, with built-in safeguards and access controls

* **Tests**
  * Comprehensive test suite added validating the new feature across multiple scenarios, including edge cases, unauthorized access attempts, and successful wrapping operations

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: src/MainnetController.sol [hunk 1]
```diff
@@ -541,6 +541,25 @@ contract MainnetController is ReentrancyGuard, AccessControlEnumerable {
...
+    /**********************************************************************************************/
+
+    function wrapAllProxyETH() external nonReentrant {
+        _checkRole(RELAYER);
+
+        uint256 proxyBalance = address(proxy).balance;
+
+        if (proxyBalance > 0) {
+            // Wrap into WETH
+            proxy.doCallWithValue(
```
Comment (supercontracts) [inline src/MainnetController.sol]: can remove this comment which is too basic

File: test/mainnet-fork/WrapETH.t.sol [hunk 2]
```diff
@@ -0,0 +1,69 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
```
Comment (supercontracts) [inline test/mainnet-fork/WrapETH.t.sol:1]: rename the file to WrapETH.t.sol

File: test/mainnet-fork/WrapEth.t.sol [hunk 3]
```diff
@@ -0,0 +1,69 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity >=0.8.0;
+
+import { ReentrancyGuard } from "../../lib/openzeppelin-contracts/contracts/utils/ReentrancyGuard.sol";
+
+import "./ForkTestBase.t.sol";
+
+contract WrapEthTestBase is ForkTestBase {
```
Comment (supercontracts) [inline test/mainnet-fork/WrapEth.t.sol]: use WrapETH... everywhere

File: test/mainnet-fork/WrapEth.t.sol [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+            "AccessControlUnauthorizedAccount(address,bytes32)",
+            address(this),
+            RELAYER
+        ));
+        mainnetController.wrapAllProxyETH();
+    }
+
+}
+
+contract WrapEthSuccessTests is WrapEthTestBase {
```
Comment (supercontracts) [inline test/mainnet-fork/WrapEth.t.sol]: same here WrapETHSuccessTests

File: test/mainnet-fork/WrapEth.t.sol [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    function test_wrapAllProxyETH_zeroBalance() external {
```
Comment (supercontracts) [inline test/mainnet-fork/WrapEth.t.sol]: need a new line above

File: test/mainnet-fork/WrapETH.t.sol [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        mainnetController.wrapAllProxyETH();
+    }
+
+}
+
+contract WrapEthSuccessTests is WrapEthTestBase {
+    function test_wrapAllProxyETH_zeroBalance() external {
+        assertEq(address(almProxy).balance, 0);
+
+        uint256 initialWethBalance = weth.balanceOf(address(almProxy));
```
Comment (supercontracts) [inline test/mainnet-fork/WrapETH.t.sol]: assert below initialWethBalance == 0

File: test/mainnet-fork/WrapETH.t.sol [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        assertEq(address(almProxy).balance,         0);
+        assertEq(weth.balanceOf(address(almProxy)), initialWethBalance);
+    }
+
+    function test_wrapAllProxyETH() external {
+        vm.deal(address(almProxy), 1 ether);
+
+        assertEq(address(almProxy).balance, 1 ether);
+
+        uint256 initialWethBalance = weth.balanceOf(address(almProxy));
```
Comment (supercontracts) [inline test/mainnet-fork/WrapETH.t.sol]: same here assert the initialWethBalance

File: src/MainnetController.sol [hunk 8]
```diff
@@ -541,6 +541,25 @@ contract MainnetController is ReentrancyGuard, AccessControlEnumerable {
...
+    /**********************************************************************************************/
+    /*** Relayer wrap ETH function                                                              ***/
+    /**********************************************************************************************/
+
+    function wrapAllProxyETH() external nonReentrant {
+        _checkRole(RELAYER);
+
+        uint256 proxyBalance = address(proxy).balance;
+
+        if (proxyBalance > 0) {
```
Comment (deluca-mike) [inline src/MainnetController.sol]: @lucas-manuel Should we no-op if there is no balance, or revert?

If we are reverting, it should be:
```solidity
require(proxyBalance > 0, "error here");

proxy.doCallWithValue...
```

If we are doing a no-op, it should be:
```solidity
if (proxyBalance == 0) return;

proxy.doCallWithValue...
```

Less indents with exit early.

File: test/mainnet-fork/WrapETH.t.sol (hunk 2 above)
Comment (deluca-mike) [reply test/mainnet-fork/WrapETH.t.sol:1]: @TheMj0ln1r you may need to make 2 commits. One to rename to `WrapEth2.t.sol` and then another to rename to `WrapETH.t.sol` since git usually doesn't see case changes in filename renames.

File: test/mainnet-fork/WrapEth.t.sol [hunk 9]
```diff
@@ -0,0 +1,69 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity >=0.8.0;
+
+import { ReentrancyGuard } from "../../lib/openzeppelin-contracts/contracts/utils/ReentrancyGuard.sol";
+
+import "./ForkTestBase.t.sol";
```
Comment (deluca-mike) [inline test/mainnet-fork/WrapEth.t.sol]: Always use explicit imports.
```suggestion
{ ForkTestBase } from import "./ForkTestBase.t.sol";
```

File: test/mainnet-fork/WrapEth.t.sol [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+        vm.prank(relayer);
+        mainnetController.wrapAllProxyETH();
+
+        _assertReentrancyGuardWrittenToTwice();
+
+        assertEq(address(almProxy).balance,         0);
+        assertEq(weth.balanceOf(address(almProxy)), initialWethBalance + 1 ether);
+    }
+
+}
```
Comment (deluca-mike) [inline test/mainnet-fork/WrapEth.t.sol]: New line at end of file. There is a vscode setting to always terminate every file with an empty line, on save.

**Review by deluca-mike — CHANGES_REQUESTED** [review]

File: src/MainnetController.sol (hunk 8 above)
Comment (TheMj0ln1r) [reply src/MainnetController.sol]: no-op early return looks good. Added.

File: test/mainnet-fork/WrapETH.t.sol (hunk 6 above)
Comment (lucas-manuel) [reply test/mainnet-fork/WrapETH.t.sol]: In both tests just inline weth.balanceOf in the assertEq

Comment (lucas-manuel) [reply test/mainnet-fork/WrapETH.t.sol]:
```solidity
    function test_wrapAllProxyETH_zeroBalance() external {
        assertEq(address(almProxy).balance,         0);
        assertEq(weth.balanceOf(address(almProxy)), 0);

        vm.record();

        vm.prank(relayer);
        mainnetController.wrapAllProxyETH();

        _assertReentrancyGuardWrittenToTwice();

        assertEq(address(almProxy).balance,         0);
        assertEq(weth.balanceOf(address(almProxy)), 0);
    }

    function test_wrapAllProxyETH() external {
        vm.deal(address(almProxy), 1 ether);

        assertEq(address(almProxy).balance,         1 ether);
        assertEq(weth.balanceOf(address(almProxy)), 0);

        vm.record();

        vm.prank(relayer);
        mainnetController.wrapAllProxyETH();

        _assertReentrancyGuardWrittenToTwice();

        assertEq(address(almProxy).balance,         0);
        assertEq(weth.balanceOf(address(almProxy)), 1 ether);
    }
```

**Review by lucas-manuel — CHANGES_REQUESTED** [review]

**Review by lucas-manuel — APPROVED** [review]
